File: src/core/encryption/encryption.js
const crypto = require('crypto');
const config = require('./config');

/**
 * Core encryption service using AES-256 in GCM mode
 * Implements zero-knowledge architecture where encryption/decryption 
 * only occurs at the device level
 */
class EncryptionService {
  /**
   * Derives a key from a password using PBKDF2
   * @param {string} password - The master password
   * @param {Buffer|null} salt - The salt (if null, a new one is generated)
   * @returns {Object} - Object containing the derived key and the salt used
   */
  async deriveKey(password, salt = null) {
    if (!salt) {
      salt = crypto.randomBytes(config.SALT_LENGTH);
    }
    
    return new Promise((resolve, reject) => {
      crypto.pbkdf2(
        password, 
        salt, 
        config.PBKDF2_ITERATIONS, 
        config.KEY_LENGTH, 
        config.PBKDF2_DIGEST, 
        (err, derivedKey) => {
          if (err) return reject(err);
          resolve({
            key: derivedKey,
            salt: salt
          });
        }
      );
    });
  }

  /**
   * Encrypts data using AES-256-GCM
   * @param {Buffer} key - The encryption key
   * @param {string|Object} data - The data to encrypt
   * @returns {Object} - The encrypted data with IV and auth tag
   */
  encrypt(key, data) {
    // Convert objects to JSON strings for encryption
    if (typeof data === 'object') {
      data = JSON.stringify(data);
    }
    
    const iv = crypto.randomBytes(config.IV_LENGTH);
    const cipher = crypto.createCipheriv(
      config.ENCRYPTION_ALGORITHM, 
      key, 
      iv, 
      { authTagLength: config.AUTH_TAG_LENGTH }
    );
    
    let encrypted = cipher.update(data, 'utf8', 'base64');
    encrypted += cipher.final('base64');
    const authTag = cipher.getAuthTag();
    
    return {
      encrypted: encrypted,
      iv: iv.toString('base64'),
      authTag: authTag.toString('base64')
    };
  }

  /**
   * Decrypts data using AES-256-GCM
   * @param {Buffer} key - The decryption key
   * @param {Object} encryptedData - Object containing encrypted data, IV, and auth tag
   * @returns {string|Object} - The decrypted data
   */
  decrypt(key, encryptedData) {
    const iv = Buffer.from(encryptedData.iv, 'base64');
    const authTag = Buffer.from(encryptedData.authTag, 'base64');
    
    const decipher = crypto.createDecipheriv(
      config.ENCRYPTION_ALGORITHM, 
      key, 
      iv, 
      { authTagLength: config.AUTH_TAG_LENGTH }
    );
    
    decipher.setAuthTag(authTag);
    
    let decrypted = decipher.update(encryptedData.encrypted, 'base64', 'utf8');
    decrypted += decipher.final('utf8');
    
    // Try to parse as JSON if it's a JSON string
    try {
      return JSON.parse(decrypted);
    } catch (e) {
      // If not a valid JSON string, return as is
      return decrypted;
    }
  }

  /**
   * Securely wipes a buffer from memory by overwriting it
   * @param {Buffer} buffer - The buffer to wipe
   */
  secureWipe(buffer) {
    if (buffer && buffer.fill) {
      // Overwrite the buffer with zeros
      buffer.fill(0);
    }
  }

  /**
   * Creates a vault key from a master password
   * @param {string} masterPassword - The master password
   * @param {Buffer|null} salt - The salt (if null, a new one is generated)
   * @returns {Object} - Object containing the vault key and the salt used
   */
  async createVaultKey(masterPassword, salt = null) {
    const { key, salt: usedSalt } = await this.deriveKey(masterPassword, salt);
    return {
      vaultKey: key,
      salt: usedSalt
    };
  }
}

module.exports = new EncryptionService(); 

File: src/crypto/encryption.py
"""
Core encryption service.
Implements AES-256-GCM encryption with zero-knowledge design.
"""

import os
import hashlib
import base64
import sys
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

# Add the project root to the Python path
sys.path.append(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))

from config.config import (
    AES_KEY_LENGTH, 
    AES_IV_LENGTH, 
    AES_AUTH_TAG_LENGTH,
    PBKDF2_ITERATIONS,
    PBKDF2_HASH_TYPE,
    PBKDF2_SALT_LENGTH
)


class EncryptionService:
    """Core encryption service using AES-256-GCM"""
    
    def derive_key(self, password, salt=None, purpose='encryption'):
        """
        Derive a key from a password using PBKDF2
        
        Args:
            password (str): The password to derive key from
            salt (str or bytes, optional): The salt for key derivation, generated if None
            purpose (str): Purpose identifier for the key
            
        Returns:
            bytes or tuple: The derived key, or tuple of (key, salt) if new salt was generated
        """
        # Ensure password is bytes
        if isinstance(password, str):
            password = password.encode('utf-8')
        
        # Generate salt if not provided
        if salt is None:
            salt = os.urandom(PBKDF2_SALT_LENGTH)
            new_salt = True
        else:
            # If salt is provided as base64, decode it
            if isinstance(salt, str):
                try:
                    salt = base64.b64decode(salt)
                except:
                    # If decoding fails, use the bytes directly
                    salt = salt.encode('utf-8')
            new_salt = False
        
        # Add purpose to prevent key reuse for different functions
        if isinstance(purpose, str):
            purpose = purpose.encode('utf-8')
        
        # Derive key using PBKDF2
        key = hashlib.pbkdf2_hmac(
            PBKDF2_HASH_TYPE,
            password + purpose,  # Add purpose to the password
            salt,
            PBKDF2_ITERATIONS,
            AES_KEY_LENGTH
        )
        
        # Return key and salt if new salt was generated
        if new_salt:
            return key, base64.b64encode(salt).decode('utf-8')
        return key
    
    def encrypt(self, key, data):
        """
        Encrypt data using AES-256-GCM
        
        Args:
            key (bytes): Encryption key (must be 32 bytes for AES-256)
            data (str or bytes): Data to encrypt
            
        Returns:
            tuple: (encrypted_data, iv, auth_tag)
        """
        # Check key length
        if len(key) != AES_KEY_LENGTH:
            raise ValueError(f"Key must be {AES_KEY_LENGTH} bytes for AES-256")
        
        # Convert data to bytes if it's a string
        if isinstance(data, str):
            data = data.encode('utf-8')
        
        # Generate a random initialization vector (IV)
        iv = os.urandom(AES_IV_LENGTH)
        
        # Create AESGCM cipher with the key
        cipher = AESGCM(key)
        
        # Encrypt the data
        # In GCM mode, the auth tag is appended to the ciphertext
        ciphertext_with_tag = cipher.encrypt(iv, data, None)
        
        # Split the ciphertext and authentication tag
        ciphertext = ciphertext_with_tag[:-AES_AUTH_TAG_LENGTH]
        auth_tag = ciphertext_with_tag[-AES_AUTH_TAG_LENGTH:]
        
        return ciphertext, iv, auth_tag
    
    def decrypt(self, key, ciphertext, iv, auth_tag):
        """
        Decrypt data using AES-256-GCM
        
        Args:
            key (bytes): Decryption key
            ciphertext (bytes): Encrypted data
            iv (bytes): Initialization vector
            auth_tag (bytes): Authentication tag
            
        Returns:
            str: Decrypted data as a string
        """
        # Check key length
        if len(key) != AES_KEY_LENGTH:
            raise ValueError(f"Key must be {AES_KEY_LENGTH} bytes for AES-256")
        
        # Ensure all inputs are bytes
        if isinstance(ciphertext, str):
            ciphertext = ciphertext.encode('utf-8')
        if isinstance(iv, str):
            iv = iv.encode('utf-8')
        if isinstance(auth_tag, str):
            auth_tag = auth_tag.encode('utf-8')
        
        # Reconstruct the full ciphertext with auth tag
        full_ciphertext = ciphertext + auth_tag
        
        # Create AESGCM cipher with the key
        cipher = AESGCM(key)
        
        # Decrypt the data
        # This will also verify the authentication tag
        plaintext = cipher.decrypt(iv, full_ciphertext, None)
        
        # Return the plaintext as a string
        return plaintext.decode('utf-8')
    
    def secure_wipe(self, buffer):
        """
        Securely wipe a buffer from memory by overwriting it
        
        Args:
            buffer (bytearray): The buffer to wipe
            
        Returns:
            None
        """
        # Check if the buffer is a bytearray or bytes object
        if isinstance(buffer, (bytearray, bytes)):
            # If it's bytes, we need to convert to bytearray
            if isinstance(buffer, bytes):
                buffer = bytearray(buffer)
            
            # Overwrite with random data
            for i in range(len(buffer)):
                buffer[i] = 0
        else:
            raise TypeError("Buffer must be a bytearray or bytes object")
    
    def create_vault_key(self, master_password):
        """
        Create a vault key from a master password
        
        Args:
            master_password (str): The master password
            
        Returns:
            tuple: (vault_key, salt)
        """
        return self.derive_key(master_password, purpose='vault_key')


# Singleton instance
encryption_service = EncryptionService() 